linux/Platform.c

```diff
@@ -538,3 +541,323 @@ bool Platform_getNetworkIO(unsigned long int* bytesReceived,
+         xSnprintf(filePath, sizeof filePath, SYS_POWERSUPPLY_DIR "/%s/uevent", entryName);
+         int fd2 = open(filePath, O_RDONLY);
+         if (fd2 == -1) {
+            closedir(dir);
+            return;
+         }
+         char buffer[1024];
+         ssize_t buflen = xread(fd2, buffer, 1023);
+         close(fd2);
+         if (buflen < 1) {
+            closedir(dir);
+            return;
+         }
+         buffer[buflen] = '\0';
+         char* buf = buffer;
+         char* line = NULL;
+         bool full = false;
+         bool now = false;
+         int fullSize = 0;
+         double capacityLevel = NAN;
+
+         #define match(str,prefix) \
+            (String_startsWith(str,prefix) ? (str) + strlen(prefix) : NULL)
+
+         while ((line = strsep(&buf, "\n")) != NULL) {
+            const char* ps = match(line, "POWER_SUPPLY_");
+            if (!ps)
+               continue;
+            const char* capacity = match(ps, "CAPACITY=");
+            if (capacity)
+               capacityLevel = atoi(capacity) / 100.0;
+            const char* energy = match(ps, "ENERGY_");
+            if (!energy)
+               energy = match(ps, "CHARGE_");
+            if (!energy)
+               continue;
+            const char* value = (!full) ? match(energy, "FULL=") : NULL;
+            if (value) {
+               fullSize = atoi(value);
+               totalFull += fullSize;
+               full = true;
+               if (now)
+                  break;
+               continue;
+            }
+            value = (!now) ? match(energy, "NOW=") : NULL;
+            if (value) {
+               totalRemain += atoi(value);
+               now = true;
+               if (full)
+                  break;
+               continue;
+            }
+         }
+
+         #undef match
+
+         if (!now && full && !isnan(capacityLevel))
+            totalRemain += (capacityLevel * fullSize);
+
+      } else if (entryName[0] == 'A') {
+         if (*isOnAC != AC_ERROR)
+            continue;
+
+         xSnprintf(filePath, sizeof filePath, SYS_POWERSUPPLY_DIR "/%s/online", entryName);
+         int fd3 = open(filePath, O_RDONLY);
+         if (fd3 == -1) {
+            closedir(dir);
+            return;
+         }
+         char buffer[2] = "";
+         for (;;) {
+            ssize_t res = read(fd3, buffer, 1);
+            if (res == -1 && errno == EINTR)
+               continue;
+            break;
+         }
+         close(fd3);
+         if (buffer[0] == '0')
+            *isOnAC = AC_ABSENT;
+         else if (buffer[0] == '1')
+            *isOnAC = AC_PRESENT;
+      }
+   }
+   closedir(dir);
+
+   *level = totalFull > 0 ? ((double) totalRemain * 100.0) / (double) totalFull : NAN;
+}
+
+static enum { BAT_PROC, BAT_SYS, BAT_ERR } Platform_Battery_method = BAT_PROC;
+
+static time_t Platform_Battery_cacheTime;
+static double Platform_Battery_cacheLevel = NAN;
+static ACPresence Platform_Battery_cacheIsOnAC;
```

Comment:
Avoid mixing variables between functions. Better put those to the top.

Reply:
Will do - thanks @BenBE